File: frame25.py
from dataclasses import dataclass
from pathlib import Path
from typing import Optional, Tuple
from datetime import datetime

import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl

import headLine
import readInput
import drawStructure


@dataclass
class Frame25Result:
    title: str
    funit: str
    lunit: str
    itp: int
    nco: int
    ndn: int
    nne: int
    nde: int
    nnod: int
    nbc: int
    nmat: int
    nsec: int
    coor: np.ndarray
    nfix: np.ndarray
    exld: np.ndarray
    fef: np.ndarray
    idbc: np.ndarray
    prop: np.ndarray
    sect: np.ndarray
    vecty: np.ndarray
    figure_path: Optional[Path]


FTYPE = ("BEAM", "PLANE TRUSS", "PLANE FRAME", "PLANE GRID", "SPACE TRUSS", "SPACE FRAME")
IPR = np.array([[1, 2, 2, 2, 3, 3],
                [2, 2, 3, 3, 3, 6]], dtype=int)


def _parse_ints_from_line(line: str, n: int) -> Tuple[int, ...]:
    toks = line.strip().replace(",", " ").split()
    if len(toks) < n:
        raise ValueError(f"Expected at least {n} numbers, got {len(toks)}: {line!r}")
    vals = []
    for k in range(n):
        x = float(toks[k])
        vals.append(int(round(x)))
    return tuple(vals)


def frame25(filename: str | Path, save_figure: bool = True) -> Frame25Result:

    start_time = datetime.now()

    filename = Path(filename)
    ipt_path = filename.with_suffix(".ipt")

    if not ipt_path.exists():
        raise FileNotFoundError(f"Input file not found: {ipt_path}")

    with ipt_path.open("r", encoding="utf-8") as f:
        title = (f.readline() or "").rstrip("\n")
        funit = (f.readline() or "").strip()
        lunit = (f.readline() or "").strip()

        ID = "*"
        headLine.headline(ID, f)
        args_line = f.readline()
        NNOD, NBC, NMAT, NSEC, ITP, NNE, IFORCE = _parse_ints_from_line(args_line, 7)

        if not (1 <= ITP <= 6):
            raise ValueError(f"ITP must be 1..6, got {ITP}")
        nco = int(IPR[0, ITP - 1])
        ndn = int(IPR[1, ITP - 1])
        nde = ndn * NNE

        COOR, NFIX, EXLD, FEF, IDBC, PROP, SECT, VECTY = readInput.readInput(
            iread=f,
            id_char=ID,
            nnod=NNOD,
            nco=nco,
            ndn=ndn,
            nbc=NBC,
            nmat=NMAT,
            nsec=NSEC,
            iforce=IFORCE,
            nde=nde,
            itp=ITP,
        )

    fmt = "k"
    fig, ax = drawStructure.drawStructure(ITP, COOR, IDBC, NBC, lunit, fmt)

    fig.tight_layout()
    out_pdf: Optional[Path] = None
    if save_figure:
        out_pdf = filename.with_name(f"{filename}-draw.pdf")
        fig.savefig(out_pdf, format="pdf", dpi=300)

    plt.close(fig)

    end_time = datetime.now()
    duration = end_time - start_time

    print(f"Start Time: {start_time}")
    print(f"End   Time: {end_time}")
    print(f"Duration  : {duration}")

    return Frame25Result(
        title=title,
        funit=funit,
        lunit=lunit,
        itp=ITP,
        nco=nco,
        ndn=ndn,
        nne=NNE,
        nde=nde,
        nnod=NNOD,
        nbc=NBC,
        nmat=NMAT,
        nsec=NSEC,
        coor=COOR,
        nfix=NFIX,
        exld=EXLD,
        fef=FEF,
        idbc=IDBC,
        prop=PROP,
        sect=SECT,
        vecty=VECTY,
        figure_path=out_pdf,
    )

File: FRAME25.m
function FRAME25(FILENAME)
%FRAME25 - A linear analysis program for framed structures
%
% Syntax:
%   FRAME25(FILENAME)
%
% Description:
%   Reads {FILENAME}.ipt as input for a framed-structure linear analysis.
%   The program parses nodes/elements/materials/sections, handles loading
%   options (concentrated vs. fixed-end forces), and can draw the geometry.
%
% Programmer:
%   Gauss Chang (R14521220)
%   Supervised by Assistant Professor Chun-Yu Ke
%   For the course: Advanced Structural Theory
%   Department of Civil Engineering, National Taiwan University
%   Term: Fall 2025  © All Rights Reserved
%
% Inputs:
%   FILENAME  - (char|string) Base name (without extension) of the input
%               file. The program expects:
%                 - {FILENAME}.ipt : input data file
%                 - {FILENAME}.dat : output listing (echoed input, nodal
%                                    displacements, member end forces)
%
% Variables (internal):
%   NNOD   = number of nodes
%   NBC    = number of Beam-column elements
%   NCO    = number of coordinates per node
%   NDN    = number of DOFs per node
%   NNE    = number of nodes per element
%   NDE    = number of DOFs per element
%   NMAT   = number of material types
%   NSEC   = number of cross-sectional types
%   IFORCE = loading flag:
%              1 -> only concentrated loads are applied
%              2 -> fixed-end forces are required
%                   (e.g., distributed loads, fabrication errors,
%                    temperature change, support settlement)
%
% Strings from input (legacy note "CHARACTERS"):
%   FUNIT  = unit of force (e.g., kN, kip)
%   LUNIT  = unit of length (e.g., mm, in)
%
% Frame type legend:
%   FRAME TYPE      ITP   NCO   NDN   (NCO and NDN are stored in array IPR)
%   BEAM             1     1     2
%   PLANAR TRUSS     2     2     2
%   PLANAR FRAME     3     2     3
%   PLANAR GRID      4     2     3
%   SPACE TRUSS      5     3     3
%   SPACE FRAME      6     3     6
%
% Notes:
%   - Place this help block immediately after the function line (no blank
%     lines in between). The first line "%FRAME25 - ..." is the H1 line
%     used by HELP/LOOKFOR/Doc and by VS Code hover.
%   - Each subfunction in this file may include its own help block using
%     the same style, directly below its own function line.
%
% See also:
%   INPUT, drawStructure, exportgraphics

FTYPE = {'BEAM'; 'PLANE TRUSS'; 'PLANE FRAME'; 'PLANE GRID'; 'SPACE TRUSS'; 'SPACE FRAME'};
IPR = [1, 2, 2, 2, 3, 3; 2, 2, 3, 3, 3, 6];

if nargin == 0
    [~,FILENAME] = fileparts(uigetfile('*.ipt','Select Input file'));
end

% Get starting time
startTime = datetime("now");

% {FILENAME}.ipt is the input data file.
% {FILENAME}.dat includes the output of the input data and the nodal 
% displacements and member end forces.

IREAD = fopen([FILENAME '.ipt'], 'r');

% Read in the problem title and the structural data
TITLE = fgets(IREAD);
FUNIT = fgets(IREAD);
LUNIT = fgets(IREAD);
ID = '*';
HEADLINE(ID,IREAD);
line = fgets(IREAD);
args = str2num(line);
temp = num2cell(args(1:7));
[NNOD, NBC, NMAT, NSEC, ITP, NNE, IFORCE] = deal(temp{:});
NCO = IPR(1, ITP);
NDN = IPR(2, ITP);
NDE = NDN * NNE;

% Read the remaining data
[COOR, NFIX, EXLD, FEF, IDBC, PROP, SECT, VECTY] = INPUT(IREAD, ID, NNOD, NCO, NDN, NBC, NMAT, NSEC, IFORCE,NDE, ITP);
FORMAT='k';
drawStructure(ITP,COOR,IDBC,NBC,LUNIT,FORMAT);
% exportgraphics(gcf, 'Problem-1-structure.pdf', 'ContentType', 'vector');
fig = gcf; ax = gca;
set(fig,'Color','w'); 
set(ax,'Color','w', 'XColor','k','YColor','k','ZColor','k');
set(findall(fig,'Type','text'),'Color','k');
exportgraphics(fig,'structure.pdf', 'ContentType','vector', 'BackgroundColor','white');
fclose(IREAD);
endTime = datetime("now");

duration = endTime - startTime;
disp(['Start Time: ', char(startTime)]);
disp(['End   Time: ', char(endTime)]);
disp(['Duration   : ', char(duration)]);

end